spec/javascripts/admin/analytics-modules/ga4-finder-setup.spec.js

```diff
@@ -0,0 +1,251 @@
+              <label for="checkboxes-4d23386d-0" class="govuk-label govuk-checkboxes__label">Red</label>
+          </div>
+      </div>
+    </fieldset>
+  `
+
+  const radios = document.createElement('div')
+  radios.innerHTML = `
+    <fieldset class="govuk-fieldset">
+      <div class="govuk-radios">
+          <input type="radio" name="radio-group" id="radio-77f4e0a5-0" value="government-gateway" class="govuk-radios__input">
+          <label for="radio-77f4e0a5-0" class="gem-c-label govuk-label govuk-radios__label">Use Government Gateway</label>
+      </div>
+    </fieldset>
+  `
+
+  const select = document.createElement('div')
+  select.innerHTML = `
+    <select>
+      <option value="1">1</option>
+      <option value="2">2</option>
+    </select>
+  `
+
+  const selectMultiple = document.querySelector('div')
+  selectMultiple.innerHTML = select.innerHTML
+  selectMultiple.querySelector('select').setAttribute('multiple', true)
+
+  const oneDate = document.createElement('div')
+  oneDate.innerHTML = `
+    <input type="text" data-ga4-change-category="update-filter one-date" />
+  `
+
+  const clearAllLink = document.createElement('div')
+  clearAllLink.setAttribute('data-ga4-filter-type', 'filter-type')
+  clearAllLink.innerHTML = `
+    <a href='/' data-ga4-change-category="clear-all-filters">
+      Link Text
+    </a>
+  `
+
+  const trackedInputs = [
+    date,
+    radios,
+    checkboxes,
+    select,
+    selectMultiple,
+    oneDate
+  ]
+
+  container.appendChild(form)
+  document.body.appendChild(container)
+
+  beforeEach(() => {
+    form.innerHTML = ''
+  })
+
+  describe('on change within tracked form', () => {
+    let trackChangeEvent
+
+    const addInputToFormAndChange = (input) => {
+      trackChangeEvent.calls.reset()
+
+      form.innerHTML = input.outerHTML
+      const field = form.querySelector('input, select')
+
+      GOVUK.analyticsGa4.analyticsModules.Ga4FinderSetup.init()
+
+      if (field.tagName === 'SELECT') {
+        field.querySelectorAll('option')[1].selected = true
+        field.dispatchEvent(new Event('change', { bubbles: true }))
+      } else {
+        field.click()
+        field.value = '123'
+        field.dispatchEvent(new Event('change', { bubbles: true }))
+      }
+    }
+
+    it('calls `trackChangeEvent` on tracked inputs', () => {
+      trackChangeEvent = spyOn(
+        GOVUK.analyticsGa4.Ga4FinderTracker,
+        'trackChangeEvent'
+      )
+
+      trackedInputs.forEach((input) => {
+        addInputToFormAndChange(input)
+        const field = form.querySelector('input, select')
+
+        expect(trackChangeEvent).toHaveBeenCalledWith(
+          new Event('change'),
+          field.dataset.ga4ChangeCategory
+        )
+
+        form.innerHTML = ''
+      })
+    })
+  })
+
+  describe('updates dataset correctly', () => {
+    it('for a clear all fields link', () => {
```

Comment:
I think I'd reverse the concepts here so that it reads better:

```js
describe('a clear all fields link', () => {
  it('updates the dataset correctly', () => {
```

Reply:
All of them 👍

Reply:
Ah OK, in that case I probably wouldn't use an `it(updates the dataset correctly)` then. I'd rewrite them to be a `describe('THE INPUT TYPE updates the dataset correctly')` since there isn't a reverse clause. Would that be correct?

Reply:
You may have to go straight for an `it` - I don't think you can write assertions inside a `describe` block. You're right though, we may not need to nest two blocks.